src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -0,0 +1,308 @@
+    start   = &_binary_cFEfs_tar_start;
+    end     = &_binary_cFEfs_tar_end;
+    return OS_FS_SUCCESS;
+}
+
+/*
+ * parses a loaded tar into fs[0]
+ */
+uint32 tar_parse()
+{
+    if(!start || !end)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(end-start-size)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(!fs[0].devname)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+
+    assert(fs[0].root->filedes==1);
+    struct fsobj *o=fs[0].root;
+    uint32 offset = 0;
+
+    while(offset+start<end){
+        //tar ends after two empty records
+        if( !(offset+start)[0] && !(offset+start)[512]) return OS_SUCCESS;
+
+        if(tar_cphdr(offset,o))     return OS_FS_ERR_DRIVE_NOT_CREATED;
+        if(file_insert(o,&fs[0]))   return OS_FS_ERR_DRIVE_NOT_CREATED;
+        offset += round_up(o->size+500);
+        if(newfile_get(&o))         return OS_FS_ERR_DRIVE_NOT_CREATED;
+    }
+
+    //tar ends before two empty records are found
+    return OS_FS_ERROR;
+}
+
+/* 
+ * Copies up to req_sz bytes from the tar file into buf
+ */
+int32 tar_read(uint32 offset, char *buf, uint32 req_sz)
+{
+    if(!start || !end || !buf)  return OS_FS_ERROR;
+    if(end-start-size)          return OS_FS_ERROR;
+    if(offset+req_sz>size)      return OS_FS_ERROR;
+    if(req_sz==0)               return OS_FS_SUCCESS;
+    
+    uint32 i;
+    for(i=0;i<req_sz;i++){
+        buf[i]=start[i+offset];
+    }
+    return OS_FS_SUCCESS;
+}
+
+/*
+ * Copies information from a tar file header to a fsobj
+ */
+uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
+{
+    //TODO: store filenames without paths
+    //TODO: more descriptive error checking
+    assert(tar_offset<size);
+    assert(file->filedes>0);
+
+    char *location = start;
+    location += tar_offset;
+
+    file -> name        = location;
+    if(*(location+strlen(location)-1)=='/'){
+        file -> type    = FSOBJ_DIR;
+        file -> size    = 0;
+    }
+    else{ 
+        file -> type    = FSOBJ_FILE;
+        file -> size    = oct_to_dec(location+124);
+        file -> data    = location+500;
+    }
+    return OS_FS_SUCCESS;
+}
+
+//this could be more performate bithack
+uint32 round_up(uint32 offset)     
+{
+    if(offset%512) return offset + (512-(offset%512));
+    return offset;
+}
+
+/******************************************************************************
+** fsobj Level Methods
+******************************************************************************/
+
+uint32 fsobj_init(struct fsobj *o)
+{
+    assert(o);
+    assert(o->filedes>0);
+	
+    o->name     = "EXAMPLE_NAME";
+	o->type     = FSOBJ_DIR;
+	o->size     = 0;
+    o->refcnt   = 0;
+	o->data     = NULL;
+	o->child    = o->parent = NULL;
+    o->next     = o->prev = NULL;
+    return OS_FS_SUCCESS;
+}
+
+//finds the next open file
+uint32 newfile_get(struct fsobj **o)
+{
+    uint32 count=0;
+    while(count<MAX_NUM_FILES && files[count].filedes){
+        count++;
+    }
```

Comment:
I can but then it ends up as `for (count == 0 ; ... ; count++) {}` which not only seems less obvious because the desired side effect is not within the loop, but it also produces a compiler warning for statement with no side effect (even though count is a side effect)